Introduce keymap for column_mode

mruwek opened the pull request:
User can now specify key-shortcuts in main topydo config file under
'column_keymap' section in form of:

`<SHORTCUT> = <ACTION>`

Two main types of action are supported:
- built-in (one of: 'home', 'end', 'up', 'down', 'postpone', 'postpone_s'
  and 'pri')
- topydo commands (aliases included). Prefixed with 'cmd'. Commands to
  call on selected todo item should contain '{}' placeholder to mark its
  place in final command call in similiar fashion as in aliases
  definitions.

postpone, postpone_s and pri shortcuts are sort of prefixes for
arguments for respective topydo commands triggered on selected todo item:
- postpone<COUNT><PERIOD> will translate to 'cmd postpone {}
  <COUNT><PERIOD>'. postpone_s will do the same but with '-s' flag
  added.
- pri<PRIORITY> will translate to 'cmd pri {} <PRIORITY>'

Default config as an example:

```ini
[column_keymap]
gg = home
G = end
j = down
k = up
pp = postpone
ps = postpone_s
pr = pri
d = cmd del {}
e = cmd edit {}
u = cmd revert
x = cmd do {}
```

- pp23d will postpone selected item by 23 days
- ps1m will postpone selected item (threshold date included) by 1 month
- prz will set priority of selected item to (Z)

bram85 commented on topydo/lib/Utils.py:
```diff
@@ -94,3 +94,20 @@ def inner():
+    """
+    Translates urwid key event to form understandable by topydo config parser.
+    """
+    key = first_to_upper(p_key)
```
This can be replaced with:

```python
key = p_key.capitalize()
```

and the `first_to_upper` function can be removed.

mruwek replied:
Oh, totally forgot about that function. Time to re-read docs for built-in types :) Fixed with recent rebase.

bram85 commented on topydo/ui/TodoListWidget.py:
```diff
@@ -101,132 +119,199 @@ def _scroll_to_bottom(self, p_size):
+        keymap, keystates = self.keymap
+
+        shortcut = self.keystate or ''
+        if len(p_key) > 1:
```
I'd prefer to see this `if` statement as part of `translate_key_to_cfg`, and let this code throw in the shortcut unconditionally.

mruwek replied:
Done in 28fd3ba61c3d7923c6fb002a2422752aa859982f

bram85 commented on topydo/lib/Utils.py:
```diff
@@ -94,3 +94,14 @@ def inner():
         return get_terminal_size.getter()
+
+
+def translate_key_to_cfg(p_key):
```
Rename `translate_key_to_cfg` to `translate_key_to_config`.

mruwek replied:
Done in 28fd3ba

bram85 commented on topydo/ui/TodoListWidget.py:
```diff
@@ -101,132 +119,199 @@ def _scroll_to_bottom(self, p_size):
+                    # action.
+                    pp_mode = keymap[self.keystate]
+                    if pp_mode in ['postpone', 'postpone_s']:
+                        if self._postpone_selected(p_key, pp_mode) is not None:
```
These two `if` statements can be combined with `and`.

mruwek replied:
After [recent change](https://github.com/bram85/topydo/pull/101/files#diff-b9959c14ca7b8926ee9d0fee31de25d4R160) I would leave it as is.

bram85 commented on topydo/ui/TodoListWidget.py:
```diff
@@ -101,132 +119,199 @@ def _scroll_to_bottom(self, p_size):
         return True
 
-    def _command_on_selected(self, p_cmd_str):
+    def _command_on_selected(self, p_cmd_str, p_execute_signal):
```
For this method and the methods below: method names should be a verb of what they do. This helps readability, it was not obvious to me what these methods _do_.

mruwek replied:
Renaming done in c0d850ac37e2cb62f87b747d021669be503762f2.

bram85 commented:
During my usage of this PR last week I somehow got stuck in a key state several times. When reading the code, the key Escape is not implemented to break the escape the key sequence at all times.

An idea would be to show user feedback in the bottom right which keys were pressed.

mruwek commented:
@bram85:
> During my usage of this PR last week I somehow got stuck in a key state several times.

Fixed in a18f9e7.

> An idea would be to show user feedback in the bottom right which keys were pressed.

Implemented also in a18f9e7.

bram85 commented on topydo/lib/Config.py:
```diff
@@ -109,11 +112,40 @@ def __init__(self, p_path=None, p_overrides=None):
+                'L': 'swap_left',
+                'R': 'swap_right',
+                '<Left>': 'prev_column',
+                '<Right>': 'next_column',
```
Make sure that `<Left>` and `<Right>` are also in _topydo.conf_.

mruwek replied:
Done in f3b18f24337adfaff2482e2ed12c80db9b75753f.

bram85 commented on topydo/lib/Config.py:
```diff
@@ -311,6 +343,32 @@ def list_format(self):
+                n = 1
+                while n < combo_len:
+                    keystates.append(''.join(combo_list[:n]))
+                    n += 1
```
This can be rewritten:

```python
from itertools import accumulate

keystates = set()

shortcuts = self.cp.items('column_keymap')
keymap_dict = dict(shortcuts)

for combo, action in shortcuts:
  # add all possible prefixes to keystates
  keystates |= set(accumulate(combo[:-1]))

mruwek replied:
Nice, but `accumulate` should operate on `combo_list` rather than `combo` to avoid creating of garbage keystates when special keys (like `<Esc>` or `<C-s>`) are used.

I would suggest:

```python
import re

from itertools import accumulate

keystates = set()

shortcuts = self.cp.items('column_keymap')
keymap_dict = dict(shortcuts)

for combo, action in shortcuts:
    # add all possible prefixes to keystates
    combo_as_list = re.split('(<[A-Z].+?>|.)', combo)[1::2]
    if len(combo_as_list) > 1:
        keystates |= set(accumulate(combo_as_list[:-1]))
```

bram85 replied:
Looks good.

mruwek replied:
Implemented in 2ece281203a1386cd019a1bfa4bc19549a239da3 and f3b18f24337adfaff2482e2ed12c80db9b75753f.

bram85 commented on topydo/lib/Config.py:
```diff
@@ -311,6 +343,32 @@ def list_format(self):
+                    n += 1
+
+            if action in ['pri', 'postpone', 'postpone_s']:
+                keystates.append(combo)
```
When changing `keystates` to a `set`, change `append` to `add`.

bram85 commented on topydo/ui/Main.py:
```diff
@@ -312,6 +333,22 @@ def _add_column(self, p_view, p_pos=None):
+        if len(p_keystate) > 0:
+            self._keystate_visible = True
+        else:
+            self._keystate_visible = False
```
Can be condensed to:

```python
self._keystate_visible = len(p_keystate) > 0
```

mruwek replied:
Done in 9a52167eacc881e9689be4e7cb68ebfd95fdfc0d.

bram85 commented on topydo/ui/TodoListWidget.py:
```diff
@@ -101,132 +120,208 @@ def _scroll_to_bottom(self, p_size):
+                # Supplied key-shortcut matches keystate and action. Save the
+                # keystate in case user will hit another key and add an action
+                # waiting for execution if user won't type anything further.
+                self.keystate = shortcut
```
Consider making self.keystate a `@property`, which setter emits `show_keystate`. That would simplify this code a bit.

mruwek replied:
I had the same thought 2 days ago, but didn't have time to implement it yet :)

mruwek replied:
Done in 9a52167eacc881e9689be4e7cb68ebfd95fdfc0d.

bram85 commented:
Functionality-wise this PR seems to be working, great job. I have some coding style remarks, after that I think this PR can be merged to `column-ui`.
